File: internal/handlers/call_handler.go
package handlers

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"
	"chat-backend/internal/websocket"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

const (
	// CallTimeoutDuration is the maximum time a call can stay in "ringing" status
	CallTimeoutDuration = 60 * time.Second // 60 seconds timeout
	// CallTimeoutCheckInterval is how often we check for timed-out calls
	CallTimeoutCheckInterval = 30 * time.Second // Check every 30 seconds
)

type CallHandler struct {
	db  *database.Database
	hub *websocket.Hub
}

func NewCallHandler(db *database.Database, hub *websocket.Hub) *CallHandler {
	return &CallHandler{db: db, hub: hub}
}

type InitiateCallRequest struct {
	Type    string   `json:"type" binding:"required"` // video, voice, group
	ChatID  string   `json:"chat_id" binding:"required"`
	Members []string `json:"members,omitempty"`
}

func (h *CallHandler) InitiateCall(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req InitiateCallRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	chatID, err := primitive.ObjectIDFromHex(req.ChatID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid chat ID"})
		return
	}

	// Load chat to verify membership and get members when not provided.
	var chat models.Chat
	if err := h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": chatID, "members": userIDObj},
	).Decode(&chat); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Chat not found or you are not a member"})
		return
	}

	// Determine call members:
	// - If client passes explicit members (e.g., group call selection), use them.
	// - Otherwise default to chat members (so direct calls include the other party).
	memberSet := make(map[primitive.ObjectID]bool)
	memberSet[userIDObj] = true
	if len(req.Members) > 0 {
		for _, memberIDStr := range req.Members {
			memberID, err := primitive.ObjectIDFromHex(memberIDStr)
			if err != nil {
				continue
			}
			memberSet[memberID] = true
		}
	} else {
		for _, m := range chat.Members {
			memberSet[m] = true
		}
	}

	members := make([]primitive.ObjectID, 0, len(memberSet))
	for id := range memberSet {
		members = append(members, id)
	}

	call := models.Call{
		ID:        primitive.NewObjectID(),
		Type:      req.Type,
		CallerID:  userIDObj,
		ChatID:    chatID,
		Members:   members,
		Status:    "ringing",
		StartedAt: time.Now(),
	}

	_, err = h.db.MongoDB.Collection("calls").InsertOne(context.Background(), call)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to initiate call"})
		return
	}

	// Broadcast call via WebSocket to call members ONLY
	// CRITICAL: Use SINGLE delivery path to prevent duplicates
	// Prefer direct send to call members; room broadcast is fallback only
	callNotification := map[string]interface{}{
		"type":       "call",
		"call_id":    call.ID.Hex(),
		"chat_id":    chatID.Hex(),
		"call_type":  call.Type,
		"caller_id":  call.CallerID.Hex(),
		"status":     call.Status,
		"message_id": primitive.NewObjectID().Hex(), // Unique message ID for deduplication
		"timestamp":  time.Now().Unix(),
	}
	callJSON, _ := json.Marshal(callNotification)

	// PRIMARY PATH: Send directly to call members (excludes caller automatically)
	log.Printf("📞 Call initiated call_id=%s caller=%s: sending to %d call members",
		call.ID.Hex(), call.CallerID.Hex(), len(members))
	for _, memberID := range members {
		if memberID == call.CallerID {
			continue // Skip caller
		}
		h.hub.SendJSONToUser(memberID, callJSON)
	}
	
	// FALLBACK PATH: Broadcast to room EXCLUDING caller to prevent caller receiving their own call event
	// This ensures users actively in chat receive the call even if not in call members list
	// CRITICAL: Exclude sender (caller) to prevent duplicate/self-call events
	h.hub.BroadcastJSONToRoomExcludingSender(chatID, call.CallerID, callJSON)

	c.JSON(http.StatusCreated, call)
}

func (h *CallHandler) AnswerCall(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	callIDStr := c.Param("call_id")
	callID, err := primitive.ObjectIDFromHex(callIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid call ID"})
		return
	}

	// Load call to get chat ID and caller ID
	var call models.Call
	err = h.db.MongoDB.Collection("calls").FindOne(
		context.Background(),
		bson.M{"_id": callID},
	).Decode(&call)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Call not found"})
		return
	}

	// Verify user is a member of the call
	isMember := false
	for _, memberID := range call.Members {
		if memberID == userIDObj {
			isMember = true
			break
		}
	}
	if !isMember {
		c.JSON(http.StatusForbidden, gin.H{"error": "You are not a member of this call"})
		return
	}

	now := time.Now()
	updateData := bson.M{
		"status":     "active",
		"answered_at": now,
		"answered_by": userIDObj,
	}
	
	_, err = h.db.MongoDB.Collection("calls").UpdateOne(
		context.Background(),
		bson.M{"_id": callID},
		bson.M{"$set": updateData},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to answer call"})
		return
	}

	// Notify caller that call was answered
	// CRITICAL: Use SINGLE delivery path to prevent duplicates
	answerNotification := map[string]interface{}{
		"type":        "call_answered",
		"call_id":     callID.Hex(),
		"chat_id":     call.ChatID.Hex(),
		"call_type":   call.Type,
		"status":      "active",
		"answered_by": userIDObj.Hex(),
		"answered_at": now,
		"message_id":  primitive.NewObjectID().Hex(), // Unique message ID for deduplication
		"timestamp":   now.Unix(),
	}
	answerJSON, _ := json.Marshal(answerNotification)
	
	// PRIMARY PATH: Send directly to caller
	log.Printf("📞 Call answered call_id=%s answered_by=%s caller=%s: sending call_answered to caller",
		callID.Hex(), userIDObj.Hex(), call.CallerID.Hex())
	h.hub.SendJSONToUser(call.CallerID, answerJSON)
	
	// FALLBACK PATH: Also broadcast to room (safe because SendJSONToUser already delivered to caller)
	h.hub.BroadcastJSONToRoom(call.ChatID, answerJSON)

	c.JSON(http.StatusOK, gin.H{"message": "Call answered"})
}

func (h *CallHandler) EndCall(c *gin.Context) {
	callIDStr := c.Param("call_id")
	callID, err := primitive.ObjectIDFromHex(callIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid call ID"})
		return
	}

	// Load call to get chat ID and members
	var call models.Call
	err = h.db.MongoDB.Collection("calls").FindOne(
		context.Background(),
		bson.M{"_id": callID},
	).Decode(&call)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Call not found"})
		return
	}

	now := time.Now()
	_, err = h.db.MongoDB.Collection("calls").UpdateOne(
		context.Background(),
		bson.M{"_id": callID},
		bson.M{"$set": bson.M{
			"status":   "ended",
			"ended_at": now,
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to end call"})
		return
	}

	// Notify all call members that the call has ended
	// CRITICAL: Use SINGLE delivery path to prevent duplicates
	endNotification := map[string]interface{}{
		"type":       "call_ended",
		"call_id":    callID.Hex(),
		"chat_id":    call.ChatID.Hex(),
		"call_type":  call.Type,
		"status":     "ended",
		"message_id": primitive.NewObjectID().Hex(),
		"timestamp":  now.Unix(),
	}
	endJSON, _ := json.Marshal(endNotification)
	
	// PRIMARY PATH: Send directly to call members
	log.Printf("📞 Call ended call_id=%s: sending to %d call members", callID.Hex(), len(call.Members))
	for _, memberID := range call.Members {
		h.hub.SendJSONToUser(memberID, endJSON)
	}
	
	// FALLBACK PATH: Also broadcast to room (safe because SendJSONToUser already delivered)
	h.hub.BroadcastJSONToRoom(call.ChatID, endJSON)

	c.JSON(http.StatusOK, gin.H{"message": "Call ended"})
}

// StartCallTimeoutChecker starts a background goroutine that periodically checks for timed-out calls
func (h *CallHandler) StartCallTimeoutChecker() {
	go func() {
		ticker := time.NewTicker(CallTimeoutCheckInterval)
		defer ticker.Stop()

		for range ticker.C {
			h.checkTimedOutCalls()
		}
	}()
	log.Println("✅ Call timeout checker started")
}

// checkTimedOutCalls finds and ends calls that have been ringing for too long
func (h *CallHandler) checkTimedOutCalls() {
	ctx := context.Background()
	timeoutThreshold := time.Now().Add(-CallTimeoutDuration)

	// Find calls that are still ringing and started before the threshold
	filter := bson.M{
		"status":    "ringing",
		"started_at": bson.M{"$lt": timeoutThreshold},
	}

	cursor, err := h.db.MongoDB.Collection("calls").Find(ctx, filter)
	if err != nil {
		log.Printf("Error finding timed-out calls: %v", err)
		return
	}
	defer cursor.Close(ctx)

	var timedOutCalls []models.Call
	if err := cursor.All(ctx, &timedOutCalls); err != nil {
		log.Printf("Error reading timed-out calls: %v", err)
		return
	}

	for _, call := range timedOutCalls {
		now := time.Now()
		_, err := h.db.MongoDB.Collection("calls").UpdateOne(
			ctx,
			bson.M{"_id": call.ID},
			bson.M{"$set": bson.M{
				"status":   "ended",
				"ended_at": now,
			}},
		)

		if err != nil {
			log.Printf("Error ending timed-out call %s: %v", call.ID.Hex(), err)
			continue
		}

		log.Printf("⏱️ Call %s timed out after %v", call.ID.Hex(), CallTimeoutDuration)

		// Notify all call members that the call timed out
		endNotification := map[string]interface{}{
			"type":       "call_ended",
			"call_id":    call.ID.Hex(),
			"chat_id":    call.ChatID.Hex(),
			"call_type":  call.Type,
			"status":     "ended",
			"reason":     "timeout",
			"message_id": primitive.NewObjectID().Hex(),
			"timestamp":  time.Now().Unix(),
		}
		endJSON, _ := json.Marshal(endNotification)

		// PRIMARY PATH: Send directly to call members
		for _, memberID := range call.Members {
			h.hub.SendJSONToUser(memberID, endJSON)
		}
		// FALLBACK PATH: Also broadcast to room
		h.hub.BroadcastJSONToRoom(call.ChatID, endJSON)
	}

	if len(timedOutCalls) > 0 {
		log.Printf("⏱️ Ended %d timed-out call(s)", len(timedOutCalls))
	}
}

File: src/handlers/callHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { randomUUID } = require('crypto');
const { getDB } = require('../database');
const { hub } = require('../websocket/hub');

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function initiateCall(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { type, chat_id, members: memberIds } = req.body || {};

    if (!chat_id) return res.status(400).json({ error: 'chat_id is required' });

    let chatObjId;
    try {
      chatObjId = new ObjectId(chat_id);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid chat_id' });
    }

    const chat = await db.collection('chats').findOne({ _id: chatObjId });
    if (!chat) return res.status(404).json({ error: 'Chat not found' });

    if (!Array.isArray(chat.members) || !chat.members.includes(userId)) {
      return res.status(403).json({ error: 'Not a member of this chat' });
    }

    // Build members set: caller + provided members + chat members
    const membersSet = new Set([userId]);
    if (Array.isArray(memberIds)) {
      for (const m of memberIds) membersSet.add(m.toString());
    }
    // Include all chat members
    for (const m of chat.members) membersSet.add(m.toString());

    const now = new Date();
    const callDoc = {
      type: type || 'voice',
      caller_id: userId,
      chat_id: chat_id.toString(),
      members: Array.from(membersSet),
      status: 'ringing',
      started_at: now,
      created_at: now,
      updated_at: now,
    };

    const insertResult = await db.collection('calls').insertOne(callDoc);
    callDoc._id = insertResult.insertedId;
    const callIdStr = insertResult.insertedId.toString();

    // Notify each member (except caller)
    for (const memberId of callDoc.members) {
      if (memberId !== userId) {
        hub.sendToUser(memberId, {
          type: 'call',
          call_id: callIdStr,
          chat_id: chat_id.toString(),
          call_type: callDoc.type,
          caller_id: userId,
          status: 'ringing',
          message_id: randomUUID(),
          timestamp: now.toISOString(),
          sender_id: userId,
        });
      }
    }

    return res.status(201).json({ ...callDoc, id: callIdStr });
  } catch (err) {
    console.error('initiateCall error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function answerCall(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const callIdStr = req.params.call_id;

    let callObjId;
    try {
      callObjId = new ObjectId(callIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid call ID' });
    }

    const call = await db.collection('calls').findOne({ _id: callObjId });
    if (!call) return res.status(404).json({ error: 'Call not found' });

    if (!Array.isArray(call.members) || !call.members.includes(userId)) {
      return res.status(403).json({ error: 'Not a member of this call' });
    }

    const now = new Date();
    await db.collection('calls').updateOne(
      { _id: callObjId },
      {
        $set: {
          status: 'active',
          answered_at: now,
          answered_by: userId,
          updated_at: now,
        },
      }
    );

    hub.sendToUser(call.caller_id, {
      type: 'call_answered',
      call_id: callIdStr,
      chat_id: call.chat_id,
      call_type: call.type,
      status: 'active',
      caller_id: call.caller_id,
      answered_by: userId,
      answered_at: now.toISOString(),
      message_id: randomUUID(),
      timestamp: now.toISOString(),
      sender_id: userId,
    });

    return res.json({ message: 'Call answered', call_id: callIdStr });
  } catch (err) {
    console.error('answerCall error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function endCall(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const callIdStr = req.params.call_id;

    let callObjId;
    try {
      callObjId = new ObjectId(callIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid call ID' });
    }

    const call = await db.collection('calls').findOne({ _id: callObjId });
    if (!call) return res.status(404).json({ error: 'Call not found' });

    const now = new Date();
    await db.collection('calls').updateOne(
      { _id: callObjId },
      {
        $set: {
          status: 'ended',
          ended_at: now,
          updated_at: now,
        },
      }
    );

    const payload = {
      type: 'call_ended',
      call_id: callIdStr,
      chat_id: call.chat_id,
      call_type: call.type,
      status: 'ended',
      message_id: randomUUID(),
      timestamp: now.toISOString(),
    };

    if (Array.isArray(call.members)) {
      for (const memberId of call.members) {
        hub.sendToUser(memberId.toString(), payload);
      }
    }

    return res.json({ message: 'Call ended', call_id: callIdStr });
  } catch (err) {
    console.error('endCall error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

function startCallTimeoutChecker() {
  setInterval(async () => {
    try {
      const db = getDB();
      const cutoff = new Date(Date.now() - 60 * 1000); // 60 seconds ago

      const timedOutCalls = await db.collection('calls').find({
        status: 'ringing',
        started_at: { $lt: cutoff },
      }).toArray();

      if (timedOutCalls.length === 0) return;

      const ids = timedOutCalls.map(c => c._id);
      const now = new Date();

      await db.collection('calls').updateMany(
        { _id: { $in: ids } },
        { $set: { status: 'ended', ended_at: now, updated_at: now } }
      );

      for (const call of timedOutCalls) {
        const callIdStr = call._id.toString();
        const payload = {
          type: 'call_ended',
          call_id: callIdStr,
          chat_id: call.chat_id,
          call_type: call.type,
          status: 'ended',
          reason: 'timeout',
          message_id: randomUUID(),
          timestamp: now.toISOString(),
        };

        if (Array.isArray(call.members)) {
          for (const memberId of call.members) {
            hub.sendToUser(memberId.toString(), payload);
          }
        }
      }

      console.log(`Call timeout checker: ended ${timedOutCalls.length} ringing call(s)`);
    } catch (err) {
      console.error('startCallTimeoutChecker error:', err);
    }
  }, 30 * 1000); // every 30 seconds
}

module.exports = {
  initiateCall,
  answerCall,
  endCall,
  startCallTimeoutChecker,
};
